### PR TITLE
Moving the mutate retry logic into BigtableDataGrpcClient.

### PR DESCRIPTION
**Pull request by sduskis**

### CODE REVIEW

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 1]
```diff
@@ -80,8 +87,39 @@ public void onCompleted() {
     }
   }
 
+  @VisibleForTesting
+  public static final Predicate<MutateRowRequest> RETRY_MUTATIONS_WITH_TIMESTAMPS =
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: nit: IS_RETRYABLE_MUTATION

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 2]
```diff
@@ -80,8 +87,39 @@ public void onCompleted() {
...
+      new Predicate<MutateRowRequest>() {
+        @Override
+        public boolean apply(@Nullable MutateRowRequest mutateRowRequest) {
+          return mutateRowRequest != null
+              && !hasMissingTimeStamp(mutateRowRequest.getMutationsList());
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: hasTimeStamp()

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 3]
```diff
@@ -80,8 +87,39 @@ public void onCompleted() {
...
+        }
+      };
+
+  @VisibleForTesting
+  public static final Predicate<CheckAndMutateRowRequest> RETRY_CHECK_AND_MUTATE_WITH_TIMESTAMP =
+      new Predicate<CheckAndMutateRowRequest>() {
+        @Override
+        public boolean apply(@Nullable CheckAndMutateRowRequest checkAndMutateRowRequest) {
+          return checkAndMutateRowRequest != null
+              && !hasMissingTimeStamp(checkAndMutateRowRequest.getTrueMutationsList())
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: I thought this should be:
checkAndMutateRowRequest != null
&&
(hasTimeStamp(checkAndMutateRowRequest.getTrueMutationsList()) || hasTimeStamp(checkAndMutateRowRequest.getFalseMutationsList()))

**Comment by sduskis** [comment]
Addressed all comments.  PTAL.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 4]
```diff
@@ -122,30 +164,45 @@ public BigtableDataGrpcClient(
...
+  }
+
+  private ClientCall<CheckAndMutateRowRequest, CheckAndMutateRowResponse>
+      createCheckAndMutateRowCall(CheckAndMutateRowRequest request) {
+    return createRetryableCall(BigtableServiceGrpc.METHOD_CHECK_AND_MUTATE_ROW,
+      IS_RETRYABLE_CHECK_AND_MUTATE, request);
+  }
+
+  private <ReqT, RespT> ClientCall<ReqT, RespT> createRetryableCall(
+      MethodDescriptor<ReqT, RespT> method, Predicate<ReqT> predicate, ReqT request) {
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: nit: Predicate<ReqT> predicate => Predicate<ReqT> isRetryable

**Comment by kevinsi4508** [comment]
LGTM

**Pull request merged**